File: src/main/java/com/godwin/jsonparser/ui/action/JBRadioAction.java
package com.godwin.jsonparser.ui.action;

import com.intellij.ide.DataManager;
import com.intellij.openapi.actionSystem.*;
import com.intellij.openapi.actionSystem.ex.CustomComponentAction;
import com.intellij.openapi.util.Key;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.util.ui.UIUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.event.ActionListener;

/**
 * Created by Godwin on 4/21/2018 12:32 PM for json.
 *
 * @author : Godwin Joseph Kurinjikattu
 */
public class JBRadioAction extends AnAction implements CustomComponentAction {
    private final ButtonGroup mButtonGroup;
    private String mActionCommand;
    private ActionListener mActionListener;
    private boolean selected;

//    public JBRadioAction(@Nullable String text) {
//        this(text, null);
//    }

    public JBRadioAction(@Nullable String text, ButtonGroup buttonGroup) {
        super(text);
        this.mButtonGroup = buttonGroup;
    }

    public JBRadioAction(String text, String actionCommand, ButtonGroup buttonGroup, ActionListener actionListener) {
        this(text, buttonGroup);
        this.mActionCommand = actionCommand;
        this.mActionListener = actionListener;
    }

    public JBRadioAction(String text, String actionCommand, ButtonGroup buttonGroup, ActionListener actionListener, boolean selected) {
        this(text, actionCommand, buttonGroup, actionListener);
        this.selected = selected;
    }

    @Override
    public @NotNull JComponent createCustomComponent(@NotNull Presentation presentation, @NotNull String place) {
        return getComponent(presentation);
    }

    @NotNull
    private JRadioButton getComponent(Presentation presentation) {
        JRadioButton jRadioButton = new JRadioButton("");
        jRadioButton.addActionListener(e -> {
            JRadioButton radioButton = (JRadioButton) e.getSource();
            ActionToolbar actionToolbar = UIUtil.getParentOfType(ActionToolbar.class, radioButton);
            DataContext dataContext = actionToolbar != null ? actionToolbar.getToolbarDataContext() : DataManager.getInstance().getDataContext(radioButton);
            JBRadioAction.this.actionPerformed(AnActionEvent.createEvent(JBRadioAction.this, dataContext, null, "unknown", ActionUiKind.NONE, null));
            System.out.println("JBRadioAction.createCustomComponent");
            if (mActionListener != null) {
                mActionListener.actionPerformed(e);
            }
        });
        presentation.putClientProperty(Key.create("selected"), selected);
        this.updateCustomComponent(jRadioButton, presentation);
        return jRadioButton;
    }

    @Override
    public void actionPerformed(@NotNull AnActionEvent anActionEvent) {
    }

    protected void updateCustomComponent(JRadioButton radioButton, Presentation presentation) {
        radioButton.setText(presentation.getText());
        radioButton.setToolTipText(presentation.getDescription());
        radioButton.setMnemonic(presentation.getMnemonic());
        radioButton.setDisplayedMnemonicIndex(presentation.getDisplayedMnemonicIndex());
        radioButton.setSelected(Boolean.TRUE.equals(presentation.getClientProperty(Key.create("selected"))));
        radioButton.setEnabled(true);
        radioButton.setVisible(true);

        if (!StringUtil.isEmpty(mActionCommand)) {
            radioButton.setActionCommand(mActionCommand);
        }
        if (mButtonGroup != null) {
            mButtonGroup.add(radioButton);
        }
    }

    @Override
    public void update(@NotNull AnActionEvent e) {
        e.getPresentation().setVisible(true);
        e.getPresentation().setEnabled(true);
    }

    @Override
    public @NotNull ActionUpdateThread getActionUpdateThread() {
        return ActionUpdateThread.BGT;
    }
}

File: src/main/kotlin/com/godwin/jsonparser/action/JBRadioAction.kt
package com.godwin.jsonparser.action

import com.godwin.jsonparser.util.Log
import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.Presentation
import com.intellij.openapi.actionSystem.ex.CustomComponentAction
import com.intellij.openapi.util.Key
import javax.swing.ButtonGroup
import javax.swing.JComponent
import javax.swing.JRadioButton

class JBRadioAction(
    text: String?,
    private val actionCommand: String?,
    private val buttonGroup: ButtonGroup?,
    private val actionListener: ((java.awt.event.ActionEvent) -> Unit)?,
    private val selected: Boolean = false
) : AnAction(text), CustomComponentAction {

    constructor(text: String?, buttonGroup: ButtonGroup?) : this(text, null, buttonGroup, null, false)

    override fun createCustomComponent(presentation: Presentation, place: String): JComponent =
        getComponent(presentation)

    private fun getComponent(presentation: Presentation): JRadioButton {
        return JRadioButton("").apply {
            buttonGroup?.add(this)
            addActionListener { e ->
                isSelected = true
                Log.i("JBRadioAction.createCustomComponent${templatePresentation.text}")
                actionListener?.invoke(e)
            }
            presentation.putClientProperty(SELECTED_KEY, selected)
            updateCustomComponent(this, presentation)
        }
    }

    override fun actionPerformed(e: AnActionEvent) {}

    private fun updateCustomComponent(radioButton: JRadioButton, presentation: Presentation) {
        radioButton.apply {
            text = presentation.text
            toolTipText = presentation.description
            mnemonic = presentation.mnemonic
            displayedMnemonicIndex = presentation.displayedMnemonicIndex
            isSelected = presentation.getClientProperty(SELECTED_KEY) == true
            isEnabled = true
            isVisible = true
            actionCommand?.let { this.actionCommand = it }
            buttonGroup?.add(this)
        }
    }

    override fun update(e: AnActionEvent) {
        e.presentation.apply {
            isVisible = true
            isEnabled = true
        }
    }

    override fun getActionUpdateThread() = ActionUpdateThread.BGT

    companion object {
        private val SELECTED_KEY = Key.create<Boolean>("selected")
    }
}
